infrastructure/async/src/main/java/tech/pegasys/teku/infrastructure/async/SafeFuture.java

```diff
@@ -122,6 +122,17 @@ public static Interruptor createInterruptor(
     return new Interruptor(interruptFuture, exceptionSupplier);
   }
 
+  /**
+   * Repeatedly run the loop until it returns false or completes exceptionally
+   *
+   * @param loopBody A supplier for generating futures to be run in succession
+   * @return A future that will complete when looping terminates
+   */
+  public static SafeFuture<Void> asyncDoWhile(ExceptionThrowingFutureSupplier<Boolean> loopBody) {
+    return SafeFuture.of(loopBody::get)
+        .thenCompose(res -> res ? asyncDoWhile(loopBody) : SafeFuture.COMPLETE);
```

Comment:
This works well with the current use cases but at some point we probably should update it to avoid recursion if the returned future is already complete - otherwise we risk getting `StackOverflowException`.  Will be quite nice to hide that complexity in here.

Reply:
good call - will follow up on this

Reply:
Fixed here: https://github.com/ConsenSys/teku/pull/3253